### ISSUE TITLE
 Build fails in environments without network connection

### ISSUE THREAD

**Issue by etungsten** [kubernetes/kubernetes#119768]
### What happened?

`make WHAT=cmd/kubelet` fails in build environments without network connection due to missing in-tree vendored dependency `go.uber.org/automaxprocs v1.5.2`.

```
$ make WHAT=cmd/kubelet
go version go1.20.6 linux/amd64
go: downloading go.uber.org/automaxprocs v1.5.2
ncpu/main.go:23:2: unrecognized import path "go.uber.org/automaxprocs": https fetch: Get "https://go.uber.org/automaxprocs?go-get=1": dial tcp: lookup go.uber.org on 127.0.1.1:53: server misbehaving
!!! [0804 15:06:39] Call tree:
!!! [0804 15:06:39]  1: /home/etung/kubernetes/hack/lib/golang.sh:843 kube::golang::setup_env(...)
!!! [0804 15:06:39]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0804 15:06:39] Call tree:
!!! [0804 15:06:39]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1

```

### What did you expect to happen?

Build to succeed with `make WHAT=cmd/kubelet` without needing to download any go dependencies.

### How can we reproduce it (as minimally and precisely as possible)?

In a network-less build environment, run `make WHAT=cmd/kubelet`.

### Kubernetes version

Building off of the v1.28.0-rc.0 tag

### Cloud provider

N/A

### OS version

<details>

```console
$ cat /etc/os-release 
NAME="Fedora Linux"
VERSION="37 (Cloud Edition)"
ID=fedora
VERSION_ID=37
VERSION_CODENAME=""
PLATFORM_ID="platform:f37"
PRETTY_NAME="Fedora Linux 37 (Cloud Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:37"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f37/system-administrators-guide/"
SUPPORT_URL="https://ask.fedoraproject.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=37
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=37
SUPPORT_END=2023-11-14
VARIANT="Cloud Edition"
VARIANT_ID=cloud

```

</details>

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by aojea** [comment]
This is a hack/tools dependency

```
~/src/kubernetes$ grep -r automaxprocs 
hack/tools/ncpu/main.go:        "go.uber.org/automaxprocs/maxprocs"
hack/tools/tools.go:    _ "go.uber.org/automaxprocs"
hack/tools/go.mod:      go.uber.org/automaxprocs v1.5.2
hack/tools/go.sum:go.uber.org/automaxprocs v1.5.2 h1:2LxUOGiR3O6tw8ui5sZa2LAaHnsviZdVOUZw4fvbnME=
hack/tools/go.sum:go.uber.org/automaxprocs v1.5.2/go.mod h1:eRbA25aqJrxAbsLO0xy5jVwPt7FQnRgjW+efnwa1WM0=
```

added here https://github.com/kubernetes/kubernetes/pull/117016 by @dims , he may have more context

**Comment by liggitt** [comment]
hack/tools aren't vendored. if you want to ensure those are available offline, `cd hack/tools && go mod download` will do that

**Comment by hswong3i** [comment]
My hack for building with OBS (https://build.opensuse.org/package/show/home:alvistack/kubernetes-kubernetes-1.28.0-rc1), see https://github.com/alvistack/kubernetes-kubernetes/commit/1028fd421d2c1d83bcdcd3e7c25755de9787e697:
```
    git clean -xdf
    pushd hack/tools && \
        go mod download && \
        go mod vendor && \
        popd
    tar zcvf ../kubernetes_1.28.0~rc1.orig.tar.gz --exclude=.git .
    debuild -uc -us
    cp kubernetes.spec ../kubernetes_1.28.0~rc1-1.spec
    mv ../kubernetes*1.28.0~rc1*.{gz,xz,spec,dsc} /osc/home\:alvistack/kubernetes-kubernetes-1.28.0-rc1/
    rm -rf ../kubernetes*1.28.0~rc1*.*
```

**Comment by sftim** [comment]
Is this actually a bug?

**Comment by liggitt** [comment]
added in https://github.com/kubernetes/kubernetes/pull/117016

I agree we probably don't want to `go install` calls in the build path... in code generation or verify tools, that's more ok, but the build target of a particular SHA should be able to execute hermetically

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/119977 to scope down when we do this to test targets, and avoid the install if the env already explicitly set GOMAXPROCS

**Comment by liggitt** [comment]
fix merged to master and head of release-1.28 branch
